sbncode/CAFMaker/FillTrue.cxx

```diff
@@ -127,6 +130,190 @@ namespace caf {
+        // for (e.g.) the SummedADC charge extraction method.
+        //
+        // Don't use this:
+        // std::vector<sim::TrackIDE> ides = bt_serv->HitToTrackIDEs(clockData, hit);
+        //
+        // Use this:
+        art::ServiceHandle<cheat::BackTrackerService> bt_serv;
+        std::vector<sim::TrackIDE> h_ides = bt_serv->ChannelToTrackIDEs(clockData, p.channel, p.start, p.end);
+        truep.h_nelec = 0.;
+        truep.h_e = 0.;
+        for (const sim::TrackIDE &ide: h_ides) {
+          truep.h_e += ide.energy;
+          truep.h_nelec += ide.numElectrons;
+        }
+
+        // Particle based truth matching
+
+        if (!chan_2_ides.count(p.channel)) { 
+          p.truth = truep;
+          continue; 
+        }
+
+        const std::vector<const sim::IDE *> &ides = chan_2_ides.at(p.channel);
+
+        // Sum energy, nelec
+        // Charge-weighted position
+        truep.p_nelec = 0.;
+        truep.p_e = 0.;
+        truep.x = 0.;
+        truep.y = 0.;
+        truep.z = 0.;
+
+        for (const sim::IDE *ide: ides) {
+          truep.x = (truep.x*truep.p_nelec + ide->x*ide->numElectrons) / (truep.p_nelec + ide->numElectrons);
+          truep.y = (truep.y*truep.p_nelec + ide->y*ide->numElectrons) / (truep.p_nelec + ide->numElectrons);
+          truep.z = (truep.z*truep.p_nelec + ide->z*ide->numElectrons) / (truep.p_nelec + ide->numElectrons);
+
+          truep.p_e += ide->energy;
+          truep.p_nelec += ide->numElectrons;
+        }
+
+        // sim::IDE's are saved after shift from SpaceCharge
+        //
+        // Correct the mean position for space charge to find the MCParticle trajectory match
+        geo::Point_t loc_sce {truep.x, truep.y, truep.z}; // perturbed by space charge
+        geo::Point_t loc_nosce; // corrected for space charge
+        if (sce && sce->EnableSimSpatialSCE()) {
+          loc_nosce = loc_sce + sce->GetCalPosOffsets(loc_sce, p.tpc);
+        }
+        else loc_nosce = loc_sce;
+
+        TVector3 loc_nosce_v(loc_nosce.x(), loc_nosce.y(), loc_nosce.z());
+        
+        // Match to MC-trajectory
+        TVector3 direction;
+        float closest_dist = -1.;
+        int traj_index = -1;
+        for (unsigned i_traj = 0; i_traj < particle.NumberTrajectoryPoints(); i_traj++) {
+          if (closest_dist < 0. || (particle.Position(i_traj).Vect() - loc_nosce_v).Mag() < closest_dist) {
+            direction = particle.Momentum(i_traj).Vect().Unit();
+            closest_dist = (particle.Position(i_traj).Vect() - loc_nosce_v).Mag();
+            traj_index = i_traj;
+          }
+        }
+
+        // residual range
+        truep.rr = 0;
+        if (traj_index >= 0) {
+          for (int i_traj = traj_index+1; i_traj < (int)particle.NumberTrajectoryPoints(); i_traj++) {
+            truep.rr += (particle.Position(i_traj).Vect() - particle.Position(i_traj-1).Vect()).Mag();
+          }
```

Comment:
It may be guessed from my comments that I am a bit concerned of the time it takes to run this.
Here we are computing a residual range for each point of the track on each of the planes.
One thing that can be done to speed up _this_ part is to pre-compute the range as function of the trajectory point.
Something like:
```cpp
std::vector<double> getPartialRange(simb::MCTrajectory const& track) {
  if (track.empty()) return {};
  std::vector<double> ranges(track.size());
  std::size_t iPoint = track.size();
  ranges[--iPoint] = 0.0;
  while(iPoint-- > 0) {
    ranges[iPoint] = ranges[iPoint + 1] + (track.Position(iPoint+1) - track.Position(iPoint)).Vect().Mag();
  }
  return ranges;
} // getPartialRange()
```
should do (then `truep.rr = partialRanges[traj_index];`).

Reply:
Fixed -- thanks for this!